File: code/java/Benchmark.java
import java.io.FileWriter;
import java.io.IOException;

public class Benchmark {
    public static void main(String[] args) {
        // Matrix sizes to test (same as in Python)
        int[] sizes = {10, 25, 50, 100, 200, 300, 400, 500, 700};

        // Number of repetitions per matrix size
        int repetitions = 7;

        // CSV file to store results (same format as Python)
        String outputFile = "code/java/data/output/results_java.csv";

        try (FileWriter writer = new FileWriter(outputFile)) {
            writer.write("Matrix size,Average time (s)\n");

            for (int n : sizes) {
                double totalTime = 0;

                for (int i = 0; i < repetitions; i++) {
                    double[][] A = Matrix.generateMatrix(n);
                    double[][] B = Matrix.generateMatrix(n);

                    long start = System.nanoTime();
                    double[][] result = Matrix.multiply(A, B);
                    long end = System.nanoTime();

                    // Convert to seconds (same as in Python)
                    double elapsedSec = (end - start) / 1e9;
                    totalTime += elapsedSec;
                }

                double avgTime = totalTime / repetitions;

                // Display results in console
                System.out.printf("Matrix %dx%d | Avg time: %.5f s%n", n, n, avgTime);

                // Save results in the CSV file
                writer.write(n + "," + avgTime + "\n");
            }

            System.out.println("\nResults saved in: " + outputFile);

        } catch (IOException e) {
            System.err.println("Error writing CSV file: " + e.getMessage());
        }
    }
}

File: code/c/benchmark.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include "matrix.c"

//  Measures the time taken to multiply two n x n matrices
double measureTime(int n) {
    struct timeval start, stop;

    double **A = generateMatrix(n);
    double **B = generateMatrix(n);
    double **C = malloc(n * sizeof(double*));
    for (int i = 0; i < n; i++)
        C[i] = malloc(n * sizeof(double));

    gettimeofday(&start, NULL);
    C = multiply(A, B, n);
    gettimeofday(&stop, NULL);

    double elapsed = (stop.tv_sec - start.tv_sec) + 1e-6 * (stop.tv_usec - start.tv_usec);

    for (int i = 0; i < n; i++) {
        free(A[i]); free(B[i]); free(C[i]);
    }
    free(A); free(B); free(C);

    return elapsed;
}

int main() {
    int sizes[] = {10, 25, 50, 100, 200, 300, 400, 500, 700};
    int repetitions = 7;
    int numSizes = sizeof(sizes) / sizeof(sizes[0]);

    system("mkdir code\\c\\data\\output 2>nul");

    FILE *file = fopen("code/c/data/output/results_c.csv", "w");
    if (file == NULL) {
        printf("Error creating the CSV file.\n");
        return 1;
    }

    fprintf(file, "Matrix size,Average time (s)\n");

    for (int i = 0; i < numSizes; i++) {
        int n = sizes[i];
        double totalTime = 0.0;

        for (int j = 0; j < repetitions; j++) {
            totalTime += measureTime(n);
        }

        double avgTime = totalTime / repetitions;
        printf("Matrix %dx%d | Avg time: %.6f s\n", n, n, avgTime);
        fprintf(file, "%d,%.6f\n", n, avgTime);
    }

    fclose(file);
    printf("\nResults saved in: code/c/data/output/results_c.csv\n");
    return 0;
}
